### PR TITLE
webhooks: add `created_by`, `updated_by` columns to `webhooks` table.

### PR DESCRIPTION
**Pull request by sashaostrikov**
Test plan:
Updated existing database tests.
SG migration up-down-up.

Closes https://github.com/sourcegraph/sourcegraph/issues/42798

Note to reviewers
_Updates_ of `updated_by` column will be implemented when [this code](https://github.com/sourcegraph/sourcegraph/pull/42774) is merged.

### CODE REVIEW

File: migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE webhooks
+    DROP COLUMN IF EXISTS created_by,
+    DROP COLUMN IF EXISTS updated_by,
+    ADD COLUMN created_by INTEGER,
+    ADD COLUMN updated_by INTEGER;
```
Comment (mrnugget) [inline migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql:5]: I kinda feel strongly about this 😄 Matches also a lot of other foreign keys we have: https://sourcegraph.sourcegraph.com/search?q=file%3Aschema.md+REFERENCES+users%28id%29&groupBy=repo

```suggestion
    ADD COLUMN created_by_user_id INTEGER,
    ADD COLUMN updated_by_user_id INTEGER;
```

I'd also make this a proper foreign key.

File: internal/database/webhooks.go [hunk 2]
```diff
@@ -72,6 +72,7 @@ func (s *webhookStore) Create(ctx context.Context, kind, urn string, secret *typ
 		urn,
 		encryptedSecret,
 		keyID,
+		actor.FromContext(ctx).UID,
```
Comment (mrnugget) [inline internal/database/webhooks.go:75]: I know this sounds nitpicky, but I'd rather have this as a separate param to the method and have the GraphQL/service layer take care of this.

File: cmd/gitserver/server/server.go [hunk 3]
```diff
@@ -631,7 +631,7 @@ func (s *Server) cloneJobConsumer(ctx context.Context, jobs <-chan *cloneJob) {
 
 			err := s.doClone(ctx, job.repo, job.dir, job.syncer, job.lock, job.remoteURL, job.options)
 			if err != nil {
-				logger.Error("failed to clone repo", log.Error(err))
+				logger.Warn("failed to clone repo", log.Error(err))
```
Comment (mrnugget) [inline cmd/gitserver/server/server.go:634]: Seems unrelated

**Review by mrnugget — APPROVED** [review]

File: internal/database/schema.json [hunk 4]
```diff
@@ -19504,6 +19517,19 @@
           "GenerationExpression": "",
           "Comment": ""
         },
+        {
+          "Name": "updated_by",
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply internal/database/schema.json:19521]: As I understand it, these fields can be used for better UX, when listing the webhooks to see how/when it was created/updated. I don't think we need to track the history of updates, though

File: cmd/gitserver/server/server.go (hunk 3 above)
Comment (sashaostrikov) [reply cmd/gitserver/server/server.go:634]: whoops, didn't see that coming, thanks for catching this!

File: migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql (hunk 1 above)
Comment (sashaostrikov) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql:5]: I agree with a name change so it is more clear.

But we don't want to make it a foreign key, because user deletion should not affect the webhook.

File: internal/database/schema.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (rrhyne) [reply internal/database/schema.json]: This is a standardization thing. Each entity should track these items to provide the people most responsible for the current settings instantly and consistently throughout the application.

File: internal/database/webhooks.go (hunk 2 above)
Comment (ryanslade) [reply internal/database/webhooks.go:75]: I agree with @mrnugget, I think getting the user out of a request should happen as close to the initiation of the request as we can get it. By the time we get down to the DB layer we should already have all the inputs and it's only concern is to persist things.

Comment (sashaostrikov) [reply internal/database/webhooks.go:75]: okay, majority wins, will add this as a parameter 👍🏻

File: migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql (hunk 1 above)
Comment (ryanslade) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: @mrnugget We're going to experiment with making the columns nullable and adding the foreign key constraint. Just want to test that when the user is deleted the column is set to null.. at least that's what we'd expect.

Comment (mrnugget) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: You can have foreign keys without `ON DELETE CASCADE`. You can have it set to `NULL`, yes.

File: internal/database/webhooks.go (hunk 2 above)
Comment (mrnugget) [reply internal/database/webhooks.go]: I'd also think passing the `*types.Webhook` itself is fine (and what we do in a lot of other stores)

File: migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql (hunk 1 above)
Comment (sashaostrikov) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: that's the plan. now the only thing is to put it into golang's int32 so it doesn't look ugly

I think about mapping `NULL` from the DB to `-1` in `types.Webhook`. Cannot really think about other way of doing that without changing the type

File: internal/database/webhooks.go (hunk 2 above)
Comment (ryanslade) [reply internal/database/webhooks.go]: I decided against that initially since then the caller doesn't know which fields they need to provide and which ones may be changed on insert.

File: migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql (hunk 1 above)
Comment (sashaostrikov) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: but the problem is that we indeed need a special user with id==0 to flag that the webhook is created programmatically...

File: internal/database/webhooks.go (hunk 2 above)
Comment (mrnugget) [reply internal/database/webhooks.go]: Yeah, I see what you mean. In practice I haven't found that to be a problem, but I get it.

`type NewWebhookInput struct` is also an option that has precedence in our codebase, I think.

File: migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql (hunk 1 above)
Comment (mrnugget) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: So you want to have 3 states? user exists, user deleted (_do_ we hard delete users? I thought we'd only soft-delete them), user is "machine". I'd add a `auto_created boolean`, to be honest. And have `UserID int` in the `types.Webhook` type. Then you can say `if !webhook.AutoCreated && webhook.UserID == 0 { // user hard-deleted }`.

Comment (sashaostrikov) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: ### Three states

Yep, I thought about the same thing.
This is not the most pretty-looking solution, especially because we would need both `auto_created` and `auto_updated` columns, but seems the only one to support 3 states and have a foreign key to `users` table at the same time.

### User deletion

Also, you've mentioned soft and hard deleting. We actually do both, but hard-deleting a user is only performed from site admin screen.
I took a look at how users are deleted (both [soft](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@202a6e7d89ef5bc7b85638fbbfd2510c236974ea/-/blob/internal/database/users.go?L513:21) and [hard](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@202a6e7d89ef5bc7b85638fbbfd2510c236974ea/-/blob/internal/database/users.go?L570)) and although we have a bunch of foreign keys from other tables, deleting a user involves a bunch of queries ran across multiple tables in one transaction.

### Which gave me another option to implement what we want

Don't add a foreign key and keep it in one column (-1 for deleted user, 0 for auto_created and >0 for user ID) and add another query to clean up `webhooks` table when the user is deleted.

WDYT?
@ryanslade @mrnugget

Comment (mrnugget) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: > Don't add a foreign key and keep it in one column (-1 for deleted user, 0 for auto_created and >0 for user ID) and add another query to clean up webhooks table when the user is deleted.

I actually feel strongly about this and that we shouldn't do it. [In _a lot_ of parts of our codebase `0` on a `*UserID` field means either "user doesn't exist" or "doesn't exist anymore" or "no user was set".](https://sourcegraph.sourcegraph.com/search?q=UserID+%3D%3D+0&groupBy=repo). That also makes it easy to handle `NULL` in the database: it'll be a `0` in the Go type. So on hard-delete of a user you can set the column to `NULL` and you get `0` in Go and you don't need to wrestle with pointers.

I'd add a second `boolean` column _once_ we have auto-creation of webhooks (that's part of milestone 4, right? https://docs.google.com/document/d/1cldGAS0fz3wqa_9Ny1zv6SHCsIrB5EsljRWPFXXoRAg/edit#heading=h.9lqrs4lkaqow))

Comment (sashaostrikov) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: That's nice to know about 0 user ID being "mapped" to "no user", we should go this way here as well.

Then what you proposed looks like the best option, I will implement it this way.

Adding an `auto_created` column can be postponed until the time we really need it, I agree 👍🏻

Comment (ryanslade) [reply migrations/frontend/1665488828_add_columns_to_webhooks_table/up.sql]: Awesome, sounds good to me 👍

**Comment by sashaostrikov** [comment]
@ryanslade @mrnugget 

Let's have another round of review, gents :)

I've rebased onto `main` because there were conflicts, but it also gave me an opportunity to add code to update the `updated_by_user_id` column values!

Changes:
- add default to `updated_by_user_id` column
- support updating of `updated_by_user_id` column
- support soft-deletion of users which will NULLify `created_by_user_id` and `updated_by_user_id` columns

File: internal/database/schema.md [hunk 5]
```diff
@@ -3080,8 +3087,12 @@ Webhooks registered in Sourcegraph instance.
 
 **code_host_urn**: URN of a code host. This column maps to external_service_id column of repo table.
 
+**created_by_user_id**: ID of a user, who created the webhook. If equals to zero, then the user does not exist (never existed or was deleted).
+
 **secret**: Secret used to decrypt webhook payload (if supported by the code host).
 
+**updated_by_user_id**: ID of a user, who updated the webhook. If equals to zero, then the user does not exist (never existed or was deleted).
```
Comment (ryanslade) [inline internal/database/schema.md:3094]: I think this comment is out of date now

File: internal/database/users.go [hunk 6]
```diff
@@ -555,6 +555,11 @@ func (u *userStore) DeleteList(ctx context.Context, ids []int32) (err error) {
 	if err := tx.Exec(ctx, sqlf.Sprintf("UPDATE registry_extensions SET deleted_at=now() WHERE deleted_at IS NULL AND publisher_user_id IN (%s)", idsCond)); err != nil {
 		return err
 	}
+	if err := tx.Exec(ctx, sqlf.Sprintf(`UPDATE webhooks
```
Comment (ryanslade) [inline internal/database/users.go:558]: Can we add a test case for this?

File: internal/database/users.go [hunk 7]
```diff
@@ -555,6 +555,11 @@ func (u *userStore) DeleteList(ctx context.Context, ids []int32) (err error) {
...
+SET created_by_user_id = CASE WHEN webhooks.created_by_user_id IN (%s) THEN NULL ELSE webhooks.created_by_user_id END,
+updated_by_user_id = CASE WHEN webhooks.updated_by_user_id IN (%s) THEN NULL ELSE webhooks.updated_by_user_id END;`, idsCond, idsCond)); err != nil {
+		return err
+	}
```
Comment (mrnugget) [inline internal/database/users.go:562]: Hmm, why do that? I wouldn't delete the reference when the user is soft-deleted. Instead we can, when querying, check whether `user.deleted_at != null`. Either with a JOIN or with return something in the GraphQL that indicates it's a deleted user. That's _why_ we soft-delete stuff :) That allows us to display in the admin UI: "Created by deleted user 1234" and helps admin make sense of the data.

Comment (mrnugget) [reply internal/database/users.go:562]: (I'd also change the test here to make sure things work as you intend them: https://github.com/sourcegraph/sourcegraph/blob/0d77cc797b2a2435bab17ede9b9f439d202e3a2b/internal/database/users_test.go#L538 -- again, I'd recommend not resetting the ID. But you can extend the `hard-delete` test to make sure the column is reset to `NULL`)

File: internal/database/webhooks.go [hunk 8]
```diff
@@ -63,7 +62,11 @@ func (s *webhookStore) Create(ctx context.Context, kind, urn string, secret *typ
 	if secret != nil {
 		encryptedSecret, keyID, err = secret.Encrypt(ctx, s.key)
 		if err != nil || (encryptedSecret == "" && keyID == "") {
-			return nil, errors.Wrap(err, "encrypting secret")
+			if err != nil {
+				return nil, errors.Wrap(err, "encrypting secret")
+			} else {
+				return nil, errors.New("empty secret or key provided")
+			}
 		}
```
Comment (mrnugget) [inline internal/database/webhooks.go:70]: Why not simplify now that the consequences are different?

```go
			if err != nil {
				return nil, errors.Wrap(err, "encrypting secret")
			}
			if (encryptedSecret == "" && keyID == "")
				return nil, errors.New("empty secret or key provided")
			}
```

File: internal/database/webhooks.go [hunk 9]
```diff
@@ -212,12 +220,16 @@ func (s *webhookStore) Update(ctx context.Context, newWebhook *types.Webhook) (*
 	if newWebhook.Secret != nil {
 		encryptedSecret, keyID, err = newWebhook.Secret.Encrypt(ctx, s.key)
 		if err != nil || (encryptedSecret == "" && keyID == "") {
-			return nil, errors.Wrap(err, "encrypting secret")
+			if err != nil {
+				return nil, errors.Wrap(err, "encrypting secret")
+			} else {
+				return nil, errors.New("empty secret or key provided")
+			}
```
Comment (mrnugget) [inline internal/database/webhooks.go:227]: Same as above

File: internal/database/webhooks_test.go [hunk 10]
```diff
@@ -96,10 +98,63 @@ func TestWebhookCreate(t *testing.T) {
...
+		webhook1, err = webhooksStore.GetByID(ctx, webhook1.ID)
+		assert.NoError(t, err)
+		assert.Equal(t, int32(0), webhook1.CreatedByUserID)
+		assert.Equal(t, int32(0), webhook1.UpdatedByUserID)
+
+		webhook2, err = webhooksStore.GetByID(ctx, webhook2.ID)
+		assert.NoError(t, err)
+		assert.Equal(t, int32(0), webhook2.CreatedByUserID)
+		assert.Equal(t, int32(0), webhook2.UpdatedByUserID)
+	})
```
Comment (mrnugget) [inline internal/database/webhooks_test.go:153]: Good test!

**Review by ryanslade — APPROVED** [review]
LGTM

File: internal/database/schema.md [hunk 11]
```diff
@@ -3080,8 +3087,12 @@ Webhooks registered in Sourcegraph instance.
 
 **code_host_urn**: URN of a code host. This column maps to external_service_id column of repo table.
 
+**created_by_user_id**: ID of a user, who created the webhook. If equals to zero, then the user does not exist (never existed or was deleted).
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply internal/database/schema.md:3090]: it just should be "if NULL, then the user does not exist"

File: internal/database/webhooks_test.go [hunk 12]
```diff
@@ -96,10 +98,63 @@ func TestWebhookCreate(t *testing.T) {
...
+
+		// Updating webhook1 by user2 and checking that updated_by_user_id is updated
+		webhook1, err = webhooksStore.Update(ctx, UID2, webhook1)
+		assert.NoError(t, err)
+		assert.Equal(t, UID2, webhook1.UpdatedByUserID)
+
+		// Soft deleting user1 should reset created_by_user_id and keep the
+		// updated_by_user_id
+		err = usersStore.Delete(ctx, UID1)
+		assert.NoError(t, err)
```
Comment (mrnugget) [inline internal/database/webhooks_test.go:132]: See above: I'd move that assertion/test to the "subject of the test" (you're testing "user deletion" here): `users_test.go`.

**Review by mrnugget — APPROVED** [review]
See comment on user deletion. Otherwise it looks good to me!

File: internal/database/users.go (hunk 6 above)
Comment (sashaostrikov) [reply internal/database/users.go:558]: there is a `TestUsers_Delete` which contains all such checks, I'll add a webhook check there as well 👍🏻

File: internal/database/users.go (hunk 7 above)
Comment (sashaostrikov) [reply internal/database/users.go:562]: > That allows us to display in the admin UI: "Created by deleted user 1234" and helps admin make sense of the data.

That's very good that you provided such an example, thanks! I'll change the soft-deletion behaviour.
